### PR TITLE
[persist] Enforce the one-batch expectation in BlobTraceUpdates

### PR DESCRIPTION
**Pull request by bkirwi**
In practice we write down just a single batch of records per part, but the code carried around a `Vec` of batches. Flattening the vec into a single batch at the edge makes a bunch of the code that manages batches simpler, and reduces the difference between the old and new encoding path.

### Motivation

Previously-unspecified refactoring.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist/src/indexed/encoding.rs [hunk 1]
```diff
@@ -765,8 +737,8 @@ mod tests {
     fn encoded_batch_sizes() {
         fn sizes(data: DataGenerator) -> usize {
             let metrics = ColumnarMetrics::disconnected();
-            let updates = data.batches().collect();
-            let updates = BlobTraceUpdates::Row(updates);
+            let updates: Vec<_> = data.batches().collect();
+            let updates = BlobTraceUpdates::Row(ColumnarRecords::concat(&updates, &metrics));
```
Comment (bkirwi) [inline src/persist/src/indexed/encoding.rs:741]: The only code we have that takes advantage of storing multiple chunks is our test code, here and one other place. This is the smallest change to keep the tests passing, but I'm not sure if it feels good or weird...

File: src/persist-client/src/fetch.rs [hunk 2]
```diff
@@ -1101,7 +1097,6 @@ where
 /// clone is very cheap.
 #[derive(Debug, Clone, Default)]
 pub(crate) struct Cursor {
-    part_idx: usize,
     idx: usize,
 }
```
Comment (bkirwi) [inline src/persist-client/src/fetch.rs:1101]: IMO it wouldn't be silly to replace `Cursor` with just a `usize` now wherever it's used... but that's a wider change that I'd prefer to kick to a future PR.

File: src/persist-client/src/internal/encoding.rs [hunk 3]
```diff
@@ -1436,7 +1436,7 @@ impl ProtoInlineBatchPart {
         Ok(BlobTraceBatchPart {
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist-client/src/internal/encoding.rs:1436]: Good callout! I couldn't find any way to rephrase it that seemed to Add Value, so I've just deleted it for now... let me know if there's something you'd rather see there instead!

File: src/persist/src/indexed/columnar/parquet.rs [hunk 4]
```diff
@@ -190,7 +182,9 @@ pub fn decode_parquet_file_kvtd(
                 let batch = batch.map_err(|e| Error::String(e.to_string()))?;
                 ret.push(decode_arrow_batch_kvtd(batch.columns(), metrics)?);
             }
-            Ok(BlobTraceUpdates::Row(ret))
+            Ok(BlobTraceUpdates::Row(ColumnarRecords::concat(
+                &ret, metrics,
+            )))
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist/src/indexed/columnar/parquet.rs:187]: Added a warn log in the case that it's not 1, but I could also convert that into a metric if that seems better!

[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/persist/src/indexed/columnar/parquet.rs:191]: `warn!` works for me!

File: src/persist-client/src/internal/encoding.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/persist-client/src/internal/encoding.rs:1427]: Deleting it works!

**Pull request merged**